Allow user configuration of the generic math family.

eerii opened the pull request:
#### 3efcbb374ccb9d7984b3905c1de6fee737c8637d
<pre>
Allow user configuration of the generic math family.
bugs.webkit.org/show_bug.cgi?id=209595

Reviewed by Carlos Garcia Campos, Michael Catanzaro, and Adrian Perez de Castro.

Adds a configuration option so that the user can specify their preferred
math font family. If it is blank (default) the fallback math font list
will be used instead.

Depends-On: #48510

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/page/SettingsBase.cpp:
(WebCore::SettingsBase::mathFontFamily const):
(WebCore::SettingsBase::setMathFontFamily):
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/platform/graphics/FontGenericFamilies.cpp:
(WebCore::FontGenericFamilies::isolatedCopy const):
(WebCore::FontGenericFamilies::isolatedCopy):
(WebCore::FontGenericFamilies::mathFontFamily const):
(WebCore::FontGenericFamilies::setMathFontFamily):
(WebCore::FontGenericFamilies::fontFamily const):
* Source/WebCore/platform/graphics/FontGenericFamilies.h:
* Source/WebKit/UIProcess/API/C/WKPreferences.cpp:
(WKPreferencesSetMathFontFamily):
(WKPreferencesCopyMathFontFamily):
* Source/WebKit/UIProcess/API/C/WKPreferencesRef.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(_WebKitSettingsPrivate::_WebKitSettingsPrivate):
(webKitSettingsSetProperty):
(webKitSettingsGetProperty):
(webkit_settings_class_init):
(webkit_settings_get_math_font_family):
(webkit_settings_set_math_font_family):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettings):

Canonical link: <a href="https://commits.webkit.org/299434@main">https://commits.webkit.org/299434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c260b27c16138baf0f5b87a76e5b1aaebf5b761

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118859 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125051 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70922 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47122 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90205 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59723 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31272 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106565 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70712 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30334 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24679 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68710 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110977 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100715 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24866 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128099 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117373 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45766 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34564 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/api/crashtests/body-window-destroy.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98864 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46132 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102785 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98644 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25098 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44091 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22096 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42331 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45636 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51314 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146071 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45101 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37573 "Found 1 new JSC binary failure: testapi, Found 18633 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Basics/typeofcombi.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48446 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46786 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -927,6 +936,19 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+     * WebKitSettings:math-font-family:
+     *
+     * The font family used as the default for content using a math font.
+     */
+    sObjProperties[PROP_MATH_FONT_FAMILY] =
```
```suggestion
     *
     * Since: 2.52
     */
    sObjProperties[PROP_MATH_FONT_FAMILY] =
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -927,6 +936,19 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+            "math-font-family",
+            _("Math font family"),
+            _("The font family used as the default for content using math font."),
+            "",
```
Can we make the default nullptr rather than an empty string? Empty strings in public APIs can be pretty confusing.

```suggestion
            nullptr,
```

Of course, make sure the lower layers can actually handle this. If not, you can expose nullptr to the public APIs and convert to empty string internally.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2457,43 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+ * Gets the #WebKitSettings:math-font-family property.
+ *
+ * Returns: The default font family used to display content marked with math font.
+ */
```
```suggestion
 *
 * Since: 2.52
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2457,43 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+ * @math_font_family: the new default math font family
+ *
+ * Set the #WebKitSettings:math-font-family property.
+ */
```
```suggestion
 *
 * Since: 2.52
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2457,43 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+/**
+ * webkit_settings_set_math_font_family:
+ * @settings: a #WebKitSettings
+ * @math_font_family: the new default math font family
```
If you change it to use nullptr instead of empty string:

```suggestion
 * @math_font_family: (nullable): the new default math font family
```

(Otherwise, we need to add a careful description explaining that empty string is allowed and default.)

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2457,43 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+ *
+ * Gets the #WebKitSettings:math-font-family property.
+ *
+ * Returns: The default font family used to display content marked with math font.
```
If you change it to use nullptr instead of empty string:

```suggestion
 * Returns: (nullable): The default font family used to display content marked with math font.
```

(Otherwise, we need to add a careful description explaining that empty string is allowed and default.)

mcatanzaro reviewed (changes requested):
I forgot to mention: we need API tests to add new API. Fortunately, that's pretty easy. Look in Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings and start by just copying the existing test for one of the other font settings. All you need to do differently is show that setting and return nullptr both work as expected, something like:

```
    // Default math font family is nullptr.
    g_assert_null(webkit_settings_get_math_font_family(settings));
    webkit_settings_set_math_font_family(settings, "sans-serif");
    g_assert_cmpstr(webkit_settings_get_math_font_family(settings), ==, "sans-serif");
    // Can be reset to default by passing nullptr again.
    webkit_settings_set_math_font_family(settings, nullptr);
    g_assert_null(webkit_settings_get_math_font_family(settings));
```

One more thing: because this adds new API, one approval is not enough to land. This requires review from a second member of @WebKit/glib-reviewers.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2459,47 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+{
+    g_return_val_if_fail(WEBKIT_IS_SETTINGS(settings), 0);
+
+    return settings->priv->mathFontFamily.data();
```
I fear this might return an empty string rather than nullptr? We'll find out from the API test.

eerii replied:
Thank you for mentioning the API tests. They caught an issue with `webkit_settings_set_math_font_family`, I had to handle the `nullptr` case explicitly.

In this case though I think `data` should already return `nullptr` if I'm understanding [this](https://searchfox.org/wubkat/source/Source/WTF/wtf/text/CStri) right. It worked when running the test locally, so let's see how it does on the CI. Thanks! :)

mcatanzaro reviewed (approved):
OK thanks, I'm approving this. We have to wait for a second approval.

nikolaszimmermann commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -443,8 +443,16 @@ FontRanges CSSFontSelector::fontRangesForFamily(const FontDescription& fontDescr
+        if (!preferredMathFamily.isEmpty() && familyName != preferredMathFamily) {
+            auto ranges = fontRangesForFamily(fontDescription, AtomString(preferredMathFamily));
+            if (!ranges.isNull())
+                return FontRanges(WTFMove(ranges), IsGenericFontFamily::Yes);
```
This one return statement and the next one uses "return FontRanges(...)", the others use "return { ... }" - I'd suggest to turn both into "return { WTFMove(ranges), IsGenericFontFamily::Yes }"..

eerii replied:
Thank you! That seems better :)

nikolaszimmermann commented on Source/WebCore/page/SettingsBase.cpp:
```diff
@@ -189,6 +189,18 @@ void SettingsBase::setPictographFontFamily(const String& family, UScriptCode scr
+
+void SettingsBase::setMathFontFamily(const String& family, UScriptCode script)
+{
+    bool changes = fontGenericFamilies().setMathFontFamily(family, script);
```
Any particular reason for using a local variable here? I'd just use "if (fontGenericFamilies().setMathFontFamily(...)".

eerii replied:
I think that is more concise too, but I did it this way to match the other `set*FontFamily` functions.

nikolaszimmermann commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2459,53 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+    if (!g_strcmp0(priv->mathFontFamily.data(), mathFontFamily))
+        return;
+
+    String mathFontFamilyString = String::fromUTF8(mathFontFamily);
```
s/string/auto/ ?

aperezdc reviewed (approved):
LGTM! Adding one more review to rubber-stamp the changes in the public API.

eerii committed:
Allow user configuration of the generic math family.
bugs.webkit.org/show_bug.cgi?id=209595

Reviewed by Carlos Garcia Campos, Michael Catanzaro, and Adrian Perez de Castro.

Adds a configuration option so that the user can specify their preferred
math font family. If it is blank (default) the fallback math font list
will be used instead.

Depends-On: #48510

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/page/SettingsBase.cpp:
(WebCore::SettingsBase::mathFontFamily const):
(WebCore::SettingsBase::setMathFontFamily):
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/platform/graphics/FontGenericFamilies.cpp:
(WebCore::FontGenericFamilies::isolatedCopy const):
(WebCore::FontGenericFamilies::isolatedCopy):
(WebCore::FontGenericFamilies::mathFontFamily const):
(WebCore::FontGenericFamilies::setMathFontFamily):
(WebCore::FontGenericFamilies::fontFamily const):
* Source/WebCore/platform/graphics/FontGenericFamilies.h:
* Source/WebKit/UIProcess/API/C/WKPreferences.cpp:
(WKPreferencesSetMathFontFamily):
(WKPreferencesCopyMathFontFamily):
* Source/WebKit/UIProcess/API/C/WKPreferencesRef.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(_WebKitSettingsPrivate::_WebKitSettingsPrivate):
(webKitSettingsSetProperty):
(webKitSettingsGetProperty):
(webkit_settings_class_init):
(webkit_settings_get_math_font_family):
(webkit_settings_set_math_font_family):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettings):

Canonical link: https://commits.webkit.org/299434@main
